Feature/alignn model

sidnb13 opened the pull request:
- Adds in implementations for the ALIGNN model (our own implementation and [LLNL's Graphite implementation](https://github.com/LLNL/graphite/tree/main/src/graphite/nn/models)).
- Implements basic functionality to load from a checkpoint file (saving is already implemented) and resume training for long jobs.
- Implements the ability to register a new transform in `common/transforms.py` via decorator and specify an arbitrary list of them to compose via the config file. These are applied during preprocessing.

sidnb13 commented:
I’m running into an issue when adding additional keyword arguments to the `base_trainer`  class `__init__ ` method.  The from_config classmethod gives me an error saying these additional kwargs do not exist in the` __init__` method when creating a cls(...) object.

@saraheisenach any clues to why this could be happening? Here’s the link to the code: [https://github.com/Fung-Lab/MatDeepLearn_dev/blob/feature/alignn-model/matdeeplearn/trainers/base_trainer.py#:~:text=scheduler%[…]),-%40staticmethod](https://github.com/Fung-Lab/MatDeepLearn_dev/blob/feature/alignn-model/matdeeplearn/trainers/base_trainer.py#:~:text=scheduler%20import%20LRScheduler-,%40registry.register_trainer(%22base%22),),-%40staticmethod)

And the error itself:

```
Traceback (most recent call last):
  File "/global/cfs/cdirs/m3641/Sidharth/MatDeepLearn_dev/scripts/main.py", line 59, in <module>
    Runner()(config)
  File "/global/cfs/cdirs/m3641/Sidharth/MatDeepLearn_dev/scripts/main.py", line 19, in __call__
    with new_trainer_context(args=args, config=config) as ctx:
  File "/global/homes/s/sidnb13/.conda/envs/matdeeplearn/lib/python3.9/contextlib.py", line 119, in __enter__
    return next(self.gen)
  File "/global/cfs/cdirs/m3641/Sidharth/MatDeepLearn_dev/matdeeplearn/common/trainer_context.py", line 42, in new_trainer_context
    trainer = trainer_cls.from_config(config)
  File "/global/cfs/cdirs/m3641/Sidharth/MatDeepLearn_dev/matdeeplearn/trainers/base_trainer.py", line 128, in from_config
    return cls(
TypeError: __init__() got an unexpected keyword argument 'save_dir'
```
Also, printing out `cls.__init__.__code__.co_varnames` excludes the save_dir  and checkpoint_dir args.

saraheisenach commented on matdeeplearn/trainers/property_trainer.py:
```diff
@@ -51,8 +51,10 @@ def train(self):
         start_epoch = self.step // len(self.train_loader)
-        for epoch in range(start_epoch, self.max_epochs):
+        
+        for epoch in range(start_epoch, start_epoch + self.max_epochs):
```
I see, maybe we can talk more in person about this? From what I understand, you're saying that essentially `max_epochs` represents the number of epochs that we want to run per task. I was under the impression that this should be for the total number of epochs over the course of the entire training, but maybe that doesn't really matter as much? If you look here, this is how OCP does it: https://github.com/Open-Catalyst-Project/ocp/blob/main/ocpmodels/trainers/energy_trainer.py#L175

It sounds like you just want a way to break up the task, right? So maybe we have an optional param called `max_checkpoint_epochs` or something like that which can be specified in the config, and if that's included then we only train for that amount. If it's not specified, then it continues to train normally.

So that way if a job stops unexpectedly, it could resume using the load_checkpoint feature without the user having to figure out how many more epochs to do in order for it to run for the total amount of time they want, if that makes sense.

sidnb13 replied:
Sounds like a good idea, I'll implement this.

saraheisenach commented on configs/config.yml:
```diff
@@ -3,15 +3,18 @@ trainer: property
 task:
   # run_mode: train
-  identifier: "my_train_job"
+  identifier: "cgcnn-test-1"
```
Can you revert this back?

saraheisenach commented on configs/config.yml:
```diff
@@ -41,7 +44,7 @@ model:
 
 optim:
-  max_epochs: 250
+  max_epochs: 10
```
Same here

saraheisenach commented on configs/config.yml:
```diff
@@ -57,7 +60,7 @@ optim:
 
 dataset:
-  processed: False # if False, need to preprocessor data and generate .pt file
+  processed: True # if False, need to preprocessor data and generate .pt file
```
Same here

saraheisenach commented on configs/config.yml:
```diff
@@ -3,15 +3,18 @@ trainer: property
   reprocess: False
-
+  # checkpoint_dir: 
+  # save_dir: 
```
You can delete this since you added it below

saraheisenach commented on configs/examples/config_alignn.yml:
```diff
@@ -0,0 +1,95 @@
+
+trainer: property
+
```
Can you please clean up these two config files and remove all of the unnecessary spacing/commented out items?

saraheisenach commented on matdeeplearn/preprocessor/processor.py:
```diff
@@ -267,7 +280,22 @@ def get_data_list(self, dict_structures, y):
+
+        if not self.otf:
+            for transform in self.transforms:
+                try:
+                    transforms_list.append(registry.get_transform_class(transform))
+                except KeyError:
+                    raise KeyError("No such transform found for '{transform}'")
```
You actually don't need this code wrapped in a try/except clause because the registry code already handles those errors, so instead you can just do:
```suggestion
                transforms_list.append(registry.get_transform_class(transform))
```

saraheisenach commented on matdeeplearn/trainers/base_trainer.py:
```diff
@@ -98,6 +103,7 @@ def from_config(cls, config):
         # TODO: figure out what configs are passed in and how they're structured
         #  (one overall config, or individual components)
 
+        # args
```
We don't need this comment anymore now that you fixed the bug

saraheisenach commented on matdeeplearn/trainers/base_trainer.py:
```diff
@@ -107,10 +113,14 @@ def from_config(cls, config):
         loss = cls._load_loss(config["optim"]["loss"])
-
         max_epochs = config["optim"]["max_epochs"]
+
+        # kwargs
```
Same here

saraheisenach reviewed (changes requested):
Thanks for making those changes! A few things:

1. There are still some comments from last week that you may have missed. I also added a couple new ones
2. Did you set up the pre-commit with the config in the repo so that it runs before you commit and is that working correctly? I noticed a couple instances where the code style had changed but you didn't add anything, so just wanted to see if that was working properly for you

sidnb13 commented on matdeeplearn/common/data.py:
```diff
@@ -58,7 +59,7 @@ def dataset_split(
 
 def get_dataset(
-    data_path, target_index: int = 0, transform_type="GetY", large_dataset=False
+    data_path, target_index: int = 0, transform_list=[], otf=False, large_dataset=False
```
Yes, that would make more sense since it would be future-proof for transforms with init parameters.

sidnb13 commented on matdeeplearn/preprocessor/processor.py:
```diff
@@ -312,7 +334,26 @@ def get_data_list(self, dict_structures, y):
+        generate_edge_features(data_list, self.edge_steps,
+                               self.r, device=self.device)
+
+        logging.info("Applying transforms...")
```
Got it, moved this into the if statement below since we apply transforms conditional on the OTF parameter.

sidnb13 commented on matdeeplearn/models/alignn.py:
```diff
@@ -0,0 +1,399 @@
+
+
+@contextlib.contextmanager
+def prof_ctx():
+    with profile(activities=[ProfilerActivity.CUDA], record_shapes=True, profile_memory=True) as prof:
+
+        yield
+
+    print(prof.key_averages().table(sort_by="cuda_memory_usage", row_limit=10))
```
I added the docstring and moved this to `helpers.py` so it can be centralized.

sidnb13 commented:
Hi Sarah, looks like I glossed over a couple of things--sorry about that.

1. I've addressed the changes you mentioned above, as well as the ones I missed from before. A couple of things: removed `common/metrics.py` since we're switching to W&B, and this was some test code I had written a few months back,
3. I was able to get the pre-commit hook reconfigured from my end, it wasn't working earlier for some reason.

saraheisenach commented on matdeeplearn/common/data.py:
```diff
@@ -78,32 +83,37 @@ def get_dataset(
+                )
+            )
+
+    print(transforms)
+    exit(0)
```
I think you accidentally left this in haha!

saraheisenach commented on matdeeplearn/models/alignn.py:
```diff
@@ -0,0 +1,410 @@
+        min_angle: float = 0.0,
+        max_angle: float = torch.acos(torch.zeros(1)).item() * 2,
+        link: Literal["identity", "log", "logit"] = "identity",
+        **kwargs
```
Just curious, for this model, are there other arguments that you haven't specified that would make us want to have `**kwargs` here? Or are all of them defined? If they are, let's just remove this to avoid confusion

sidnb13 replied:
Yes we can remove this; I think this was left over from passing in `data` as an argument, which I believe is now not an issue.

saraheisenach commented on matdeeplearn/models/alignn_graphite.py:
```diff
@@ -0,0 +1,155 @@
+    Reference: https://www.nature.com/articles/s41524-021-00650-1.
+    """
+
+    def __init__(self, dim=64, num_interactions=4, num_species=3, cutoff=3.0, **kwargs):
```
See my comment above in the ALIGNN model regarding `kwargs`

saraheisenach commented on matdeeplearn/preprocessor/helpers.py:
```diff
@@ -1,14 +1,29 @@
+
+        yield
+
+    print(prof.key_averages().table(sort_by="cuda_memory_usage", row_limit=10))
```
Should we switch this to logging too, or does print make more sense here?

sidnb13 replied:
I think debug would make the most sense as it's a debug tool. For future development, there are better tracing tools for PyTorch such as one provided by W&B that we could look into.

saraheisenach commented on matdeeplearn/trainers/property_trainer.py:
```diff
@@ -37,22 +40,33 @@ def __init__(
         start_epoch = self.step // len(self.train_loader)
-        for epoch in range(start_epoch, self.max_epochs):
+
+        for epoch in range(start_epoch, epochs_to_run):
```
So now it should be this, right? That way if you load from a checkpoint you're running `x` epochs more starting from the epoch you're currently at? Double check me on this, though
```suggestion
        for epoch in range(start_epoch, start_epoch + epochs_to_run):
```

sidnb13 replied:
However, this would conflict with the idea of `max_epochs` if we decide to load from config but forget to specify the `max_checkpoint_epochs`. Instead, could we do the following to keep the loop unchanged?

```python
epochs_to_run = (
            self.max_checkpoint_epochs + start_epoch
            if self.max_checkpoint_epochs
            else self.max_epochs
        )
```

saraheisenach replied:
Sure, I like that. How about you change the variable name though to `end_epoch`, because `epochs_to_run` implies that I want to run 100 epochs, but really we may want to be running epoch 200-299.

saraheisenach reviewed (changes requested):
Great job! I only see a few small bug fixes and then a couple places where we can remove some unnecessary parameters

saraheisenach commented on configs/config.yml:
```diff
@@ -57,33 +53,27 @@ optim:
-  #Path to save processed data.pt file
+  # Path to save processed data.pt file
   pt_path: "/global/homes/s/shuyijia/datasets/MP_data_npj/"
-  #Format of data files (limit to those supported by ASE)
```
Could you also add in the transforms into here? Ideally we'd like to have all possible params listed in config.yml and then only the necessary ones in the examples

sidnb13 replied:
Sure, this would make things a bit more streamlined. Also in that case, we should not make `GetY` a special exception, since I had it like that before. That way it is always specified in the config to make it clear what transforms are being applied. This should eliminate use of the `target_index` parameter since it's only used for `GetY`. What do you think?

sidnb13 replied:
For transforms, I also added two modifications:
1. It looks like this was a bug which was not caught earlier. Earlier, the composition of transforms was applied to each data object inside the main loop iterating over `sdict` objects. However, node and edge features are applied after this, making this error-prone. I've modified the code to apply transforms after all preprocessing occurs in a separate loop.
2. Instead of making `otf` a global attribute used to determine on-the-fly application of transforms, I made it a transform-level parameter, so certain transforms can be applied on-the-fly by choice. This allows us to make distinguishments between things like `GetY` and `LineGraphMod`, the former of which needs to be applied during preprocessing.

saraheisenach replied:
Nice! And yeah I completely agree with including GetY so that it's clear what transforms are applied. I kind of think this also gets to a larger issue with our approach, which is that GetY isn't always needed/wanted. Take the DOS model for example, that doesn't use `y` as the target attribute at all, so it doesn't really make sense to always be doing it. Right now, though, our code will fail if `y` isn't included because we use it, at a minimum, to determine certain properties. I think in the future, we might want to rethink how we approach this, and maybe instead just have a 'target' attribute that could be `y` or anything else that the user wants.

Right now, though, if it's not already in there, we should add something to automatically fail the build if `GetY` isn't included in the config. Is that there already?

sidnb13 replied:
Sure, I could add a simple assertion at the two different areas where transforms are applied (so in `data.py` and `preprocessor.py`) that fixes this for now.

saraheisenach commented on matdeeplearn/trainers/base_trainer.py:
```diff
@@ -107,10 +110,13 @@ def from_config(cls, config):
         identifier = config["task"].get("identifier", None)
         verbosity = config["task"].get("verbosity", None)
+        # pass in custom results home dir and load in prev checkpoint dir
+        save_dir = config["task"].get("save_dir", os.getcwd())
```
Also actually why don't you switch this to be 
```suggestion
        save_dir = config["task"].get("save_dir", None)
```
because you'll need to duplicate this in the base __init__ because we don't want `save_dir` to be mandatory, but default behavior should be `os.getcwd()`

sidnb13 replied:
Actually the intention here was to replace the original `run_dir` variable from previous versions, and `save_dir` can stay optional in the config but default to `os.getcwd()`. However, it would make more sense to have `save_dir` in that case be positional in the BaseTrainer `__init__` since it's guaranteed to have a value. Could we do this instead?

sidnb13 commented:
A note on GetY dimensionality bug for documentation purposes
- When the `GetY` transform is performed on a graph, `data.y` becomes a 0D tensor, which the DOS and CGCNN models use as a fact to calculate the output dimension.
- Refactoring the `GetY` transform as something required to be specified in config introduces a new bug, it is not applied "on the fly" in `data.py` when `get_dataset` is called (unless specifed), rather, the `get_dataset` fetches the processed data but implicitly converts this 0D tensor which is created in `processor.py` into a 1D tensor, causing issues for DOS and CGCNN. The solution is to modify the line `self.output_dim = len(data[0][self.target_attr][0])` to `self.output_dim = len(data[0][self.target_attr])`.

saraheisenach commented on matdeeplearn/preprocessor/processor.py:
```diff
@@ -94,6 +97,12 @@ def __init__(
                 step size for creating Gaussian basis for edges
                 used in torch.linspace
 
+            otf: bool
+                default False. Whether or not to transform the data on the fly.
```
We should remove this now since you changed it in the config

saraheisenach commented on matdeeplearn/preprocessor/processor.py:
```diff
@@ -314,6 +328,23 @@ def get_data_list(self, dict_structures, y):
+        logging.debug("Applying transforms.")
+        transforms_list = []
+        for transform in self.transforms:
+            if not transform["otf"]:
```
Same here, let's do:
```suggestion
            if not transform.get("otf", False):
```

saraheisenach commented on matdeeplearn/common/data.py:
```diff
@@ -71,39 +74,35 @@ def get_dataset(
-    else:
-        raise ValueError("No such transform found for {transform}")
+    for transform in transform_list:
+        if transform["otf"]:
```
Let's make this optional, so default behavior is False
```suggestion
        if transform.get("otf", False):
```

saraheisenach commented on configs/config.yml:
```diff
@@ -57,33 +53,31 @@ optim:
+    - name: GetY
+      args:
+        index: 0
+      otf: False
```
Can you just add a comment next to this argument indicating that it's optional? In the future we need to find a way to keep track of optional vs required arguments in the yaml, maybe adding documentation for this

saraheisenach commented on configs/examples/config_hive.yml:
```diff
@@ -0,0 +1,83 @@
+trainer: property
+
```
Is this file something you think we should have as an example or do you think we can remove it since it seems to be the same as the CGCNN one except for the dataset paths.

sidnb13 replied:
It was a temporary file carried over from working on HIVE--I can remove it.

saraheisenach commented on configs/examples/config_test.yml:
```diff
@@ -0,0 +1,83 @@
+trainer: property
```
Same here

saraheisenach commented on matdeeplearn/common/data.py:
```diff
@@ -71,39 +74,35 @@ def get_dataset(
         Dataset = StructureDataset
 
-    transform = T(index=target_index)
+    composition = Compose(transforms) if len(transforms) > 0 else None
```
See the comment on `GetY` above. If we always default to having at least `GetY`, then transforms should always be at least length 1

sidnb13 replied:
Yes, that makes sense. And instead of defaulting to `None`, we should use `transforms[0]`.

saraheisenach commented on matdeeplearn/common/data.py:
```diff
@@ -58,7 +59,9 @@ def dataset_split(
 def get_dataset(
-    data_path, target_index: int = 0, transform_type="GetY", large_dataset=False
+    data_path,
+    transform_list: list = [],
```
Do we want this to be
```suggestion
    transform_list: list = ["GetY"],
```

This goes back to the comment where we were discussing the GetY behavior. I am under the impression that if GetY isn't included as a transform, somewhere within our code, there will be a failure because `y` is used. If this is the case, do we want to throw an error here if GetY isn't included? I think in the future it would make sense to change how we approach `y` vs other targets, but for now, we are dependent on using `y`, so I think it would fail, but double check, because maybe I'm wrong

sidnb13 replied:
I think it should remain `transform_list: List[dict] = []`. Since we instantiate GetY with an argument, and possibly more in the future, the assertion makes more sense for now as the user should specify GetY.

saraheisenach commented on matdeeplearn/preprocessor/processor.py:
```diff
@@ -41,13 +42,14 @@ def process_data(dataset_config):
         r=cutoff_radius,
         n_neighbors=n_neighbors,
         edge_steps=edge_steps,
+        transforms=dataset_config.get("transforms", []),
```
See my other comment above about maybe making the default list to be this
```suggestion
        transforms=dataset_config.get("transforms", ["GetY"]),
```

sidnb13 replied:
Same as above.

saraheisenach commented on matdeeplearn/tasks/task.py:
```diff
@@ -14,9 +13,11 @@ def __init__(self, config):
-        checkpoint = self.config.get("checkpoint", None)
-        if checkpoint is not None:
-            self.trainer.load_checkpoint(self.config["checkpoint"])
+        use_checkpoint = self.config["model"]["load_model"]
```
Let's make this optional too, so that it's only needed if a model wants to be loaded
```suggestion
        use_checkpoint = self.config["model"].get("load_model", False)
```

saraheisenach commented on matdeeplearn/trainers/base_trainer.py:
```diff
@@ -123,17 +129,22 @@ def from_config(cls, config):
-        dataset = get_dataset(dataset_path, target_index)
+        dataset = get_dataset(
+            dataset_path,
+            transform_list=dataset_config.get("transforms", []),
```
If you decide that `GetY` is needed in order to avoid errors, then I think this should instead be
```suggestion
            transform_list=dataset_config.get("transforms", ["GetY"]),
```

saraheisenach commented on matdeeplearn/trainers/property_trainer.py:
```diff
@@ -37,22 +40,32 @@ def __init__(
             logging.info("Starting regular training")
             logging.info(
-                f"running for  {self.max_epochs} epochs on {type(self.model).__name__} model"
+                f"running for {epochs_to_run} epochs on {type(self.model).__name__} model"
```
If you make the change I suggested in the comment below, then this should be
```suggestion
                f"running for {end_epoch - start_epoch} epochs on {type(self.model).__name__} model"
```
